### PR TITLE
More closely match MRI behaviour for blocks with rb_funcall and rb_iterate

### PR DESCRIPTION
**Pull request by aardvark179**

### CODE REVIEW

File: truffleruby/src/main/ruby/core/truffle/cext.rb [hunk 1]
```diff
@@ -1550,14 +1561,32 @@ def rb_thread_call_without_gvl(function, data1, unblock, data2)
...
-    call_c_with_block function, arg1 do |block_arg|
-      Truffle::Interop.execute block, block_arg, arg2
+  def rb_iterate_call_block( iter_block, block_arg, arg2)
+    Truffle::Interop.execute iter_block, block_arg, arg2
+  end
+
+  def rb_iterate(function, arg1, iter_block, arg2, block)
+    if block
+      call_c_with_block function, arg1 do |block_arg|
+        rb_iterate_call_block( iter_block, block_arg, arg2) do |*args|
```
Comment (chrisseaton) [inline truffleruby/src/main/ruby/core/truffle/cext.rb]: No space after the `(`.

File: truffleruby/src/main/ruby/core/truffle/cext.rb [hunk 2]
```diff
@@ -958,7 +958,22 @@ def rb_funcallv_public(recv, meth, args)
...
 
   def rb_funcallv(recv, meth, args)
-    recv.send(meth, *args)
+    rb_funcall(recv, meth, *args)
+  end
+
+  def rb_funcall(recv, meth, *args)
+    old_c_block = Thread.current[:__C_BLOCK__]
+    begin
+      block = Thread.current[:__C_BLOCK__]
```
Comment (eregon) [inline truffleruby/src/main/ruby/core/truffle/cext.rb]: This logic looks weird, reading twice a thread-local variable is just gonna get the same result, isn't it?

File: truffleruby/src/main/ruby/core/truffle/cext.rb [hunk 3]
```diff
@@ -1550,14 +1561,32 @@ def rb_thread_call_without_gvl(function, data1, unblock, data2)
     Thread.current.unblock unblocker, runner
   end
 
-  def rb_iterate(function, arg1, block, arg2)
-    call_c_with_block function, arg1 do |block_arg|
-      Truffle::Interop.execute block, block_arg, arg2
+  def rb_iterate_call_block( iter_block, block_arg, arg2)
```
Comment (eregon) [inline truffleruby/src/main/ruby/core/truffle/cext.rb]: Wrong spacing

File: truffleruby/src/main/ruby/core/truffle/cext.rb [hunk 4]
```diff
@@ -1550,14 +1561,32 @@ def rb_thread_call_without_gvl(function, data1, unblock, data2)
...
+    Truffle::Interop.execute iter_block, block_arg, arg2
+  end
+
+  def rb_iterate(function, arg1, iter_block, arg2, block)
+    if block
+      call_c_with_block function, arg1 do |block_arg|
```
Comment (eregon) [inline truffleruby/src/main/ruby/core/truffle/cext.rb]: Can you add some parens here to clarify?

File: truffleruby/src/main/ruby/core/truffle/cext.rb (hunk 2 above)
Comment (eregon) [inline truffleruby/src/main/ruby/core/truffle/cext.rb]: The 2 variables are essentially just aliases of each other, so only one is needed.

File: truffleruby/src/main/ruby/core/truffle/cext.rb [hunk 5]
```diff
@@ -958,7 +958,22 @@ def rb_funcallv_public(recv, meth, args)
...
+
+  def rb_funcall(recv, meth, *args)
+    old_c_block = Thread.current[:__C_BLOCK__]
+    begin
+      block = Thread.current[:__C_BLOCK__]
+      Thread.current[:__C_BLOCK__] = nil
+      if block
+        recv.send(meth, *args, &block)
+      else
+        recv.send(meth, *args)
```
Comment (eregon) [inline truffleruby/src/main/ruby/core/truffle/cext.rb]: If block is only a Proc or `nil` then you can pass it with `&block` just as well.

**Pull request merged**